operator/pkg/gateway-api/gateway_reconcile.go

```
@@ -15,7 +30,293 @@ import (
+	svcList := &corev1.ServiceList{}
+	if err := r.Client.List(ctx, svcList, client.MatchingLabels{
+		owningGatewayLabel: gw.GetName(),
+	}); err != nil {
+		return err
+	}
+
+	if len(svcList.Items) == 0 {
+		return fmt.Errorf("no service found")
+	}
+
+	svc := svcList.Items[0]
+	if len(svc.Status.LoadBalancer.Ingress) == 0 {
+		return fmt.Errorf("load balancer status is not ready")
+	}
+
+	var addresses []gatewayv1beta1.GatewayAddress
+	for _, s := range svc.Status.LoadBalancer.Ingress {
+		if len(s.IP) != 0 {
+			addresses = append(addresses, gatewayv1beta1.GatewayAddress{
+				Type:  GatewayAddressTypePtr(gatewayv1beta1.IPAddressType),
+				Value: s.IP,
+			})
+		}
+		if len(s.Hostname) != 0 {
+			addresses = append(addresses, gatewayv1beta1.GatewayAddress{
+				Type:  GatewayAddressTypePtr(gatewayv1beta1.HostnameAddressType),
+				Value: s.Hostname,
+			})
+		}
+	}
+
+	gw.Status.Addresses = addresses
+	return nil
+}
+
+func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1beta1.Gateway) error {
+	for _, l := range gw.Spec.Listeners {
+		cond := gatewayListenerReadyCondition(gw, true, "Listener Ready")
+		if l.TLS != nil {
+			for _, cert := range l.TLS.CertificateRefs {
+				if !IsSecret(cert) {
+					continue
+				}
+				secret := &corev1.Secret{}
+				if err := r.Client.Get(ctx, client.ObjectKey{
+					Namespace: namespaceDerefOr(cert.Namespace, gw.GetNamespace()),
+					Name:      string(cert.Name),
+				}, secret); err != nil {
+					if k8serrors.IsNotFound(err) {
+						cond = metav1.Condition{
+							Type:               string(gatewayv1beta1.ListenerConditionResolvedRefs),
+							Status:             metav1.ConditionFalse,
+							Reason:             string(gatewayv1beta1.ListenerReasonInvalidCertificateRef),
+							Message:            err.Error(),
+							LastTransitionTime: metav1.Now(),
+						}
+						break
+					}
+					return err
+				}
+			}
```

Is it possible that all certificate refs will fail the `IsSecret` check? If so, shouldn't we return an error after the `for` loop?

---

I was thinking about this, but as long as we set individual status for each invalid reference, the reconciliation loop is still considered as success. 

Later, if the secret is created, new reconciliation will be triggered.